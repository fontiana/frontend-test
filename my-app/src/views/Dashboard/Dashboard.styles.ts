import styled from "styled-components";
import { Input as _Input } from "../../components/Input";

export const Input = styled(_Input)``;

export const SubTitle = styled.h3`
  margin-right: 4px;
  font-size: 20px;
  margin: 12px 0;
`;

export const Section = styled.div`
  width: 100%;
  @media screen and (max-width: 500px) {
    margin-top: 12px;
  }
`;

export const Wrapper = styled.div`
  display: flex;

  @media screen and (max-width: 500px) {
    display: block;
  }
`;
