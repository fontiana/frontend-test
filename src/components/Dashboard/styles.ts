import styled from "styled-components";

export const DropdownContainer = styled.div`
  position: relative;
  display: inline-block;
`;

export const Select = styled.select`
  padding: 10px;
`;

export const Option = styled.option`
  padding: 10px;
  background-color: white;
`;

export const SelectActive = styled(Select)`
  background-color: #f5f5f5;
`;

export const CustomTable = styled.table`
  width: 100%;
  border-collapse: collapse;
  border-spacing: 0;
  background-color: #fff;
  border: 1px solid #ddd;
  font-size: 16px;
`;

export const TableHead = styled.thead`
  background-color: #f2f2f2;
`;

export const TableRow = styled.tr`
  &:nth-child(even) {
    background-color: #f2f2f2;
  }
`;

export const TableHeader = styled.th`
  padding: 10px;
  text-align: left;
`;

export const TableCell = styled.td`
  padding: 10px;
  text-align: left;
`;
