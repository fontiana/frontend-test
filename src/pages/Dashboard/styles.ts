import styled from "styled-components";
import media from "../../utils/media";

export const Wrapper = styled.main`
  display: flex;
  flex-direction: column;
  gap: 24px;

  ${media.greaterThan("tablet")`
    flex-direction: row;
  `}
`;
