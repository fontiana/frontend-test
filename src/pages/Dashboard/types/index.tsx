export interface SymbolInfoI {
  A: string;
  B: string;
  C: number;
  E: number;
  F: number;
  L: number;
  O: number;
  P: string;
  Q: string;
  a: string;
  b: string;
  c: string;
  e: string;
  h: string;
  l: string;
  n: number;
  o: string;
  p: string;
  q: string;
  s: string;
  v: string;
  w: string;
  x: string;
}

export interface ExchangeInfoI {
  [key: string]: SymbolInfoI;
}
