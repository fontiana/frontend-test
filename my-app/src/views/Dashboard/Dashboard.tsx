/* eslint-disable @typescript-eslint/no-unused-vars */
import { useQuery } from "react-query";
import { Loading } from "../../components/Loading";
import StickyHeadTable, { Column } from "../../components/Table/TableGrid";
import * as S from "./Dashboard.styles";
import { useMemo, useState } from "react";

interface ITicker {
  stream: string;
  data: {
    c: string;
    b: string;
    a: string;
    P: string;
    s: string;
  };
}

export const Dashboard = () => {
  const [symbolsState, setSymbolsState] = useState([]);
  const [symbol, setSymbol] = useState("");
  const [bidData, setBidData] = useState<ITicker>();

  const { data, error, isLoading } = useQuery("exchangeInfo", async () => {
    const response = await fetch(
      "https://data.binance.com/api/v3/exchangeInfo"
    );
    if (!response.ok) {
      throw new Error("Não foi possível obter os dados da exchange.");
    }
    const jsonData = await response.json();
    return jsonData;
  });

  const ws = new WebSocket(
    `wss:data-stream.binance.com/stream?streams=${symbol.toLowerCase()}@ticker`
  );

  ws.onmessage = (event) => {
    console.log(event.data);
    !!event.data && setBidData(event.data);
  };

  console.log(bidData, "aqui");

  const symbols = data?.symbols;

  const handleSymbolRowClick = (symbolName: string) => {
    const selectedSymbolObject = symbols?.filter(
      (item: { symbol: string }) => item.symbol === symbolName
    );
    setSymbol(selectedSymbolObject[0].symbol);
  };

  useMemo(() => {
    setSymbolsState(symbols);
  }, [symbols]);

  if (isLoading) {
    return (
      <div
        style={{
          display: "flex",
          justifyContent: "center",
          alignItems: "center",
          fontSize: "50px",
          height: "100%",
        }}
      >
        <Loading />
      </div>
    );
  }

  if (error) {
    return `Ocorreu um erro ao buscar os dados da exchange: ${error}.`;
  }

  const columns: Column[] = [
    { id: "symbol", label: "Symbol" },
    { id: "bid_price", label: "Bid Price" },
    { id: "last_price", label: "Last Price" },
    { id: "ask_price", label: "Ask Price" },
    { id: "price_change", label: "Price Change" },
  ];

  const columns1 = [{ id: "symbol", label: "Symbol" }];

  return (
    <S.Wrapper>
      <div style={{ marginRight: "20px" }}>
        <div style={{ marginBottom: "12px" }}>
          <S.Input
            placeholder="Search for a symbol..."
            onChange={(e) => {
              // eslint-disable-next-line @typescript-eslint/no-explicit-any
              setSymbolsState(() => {
                if (isLoading) return "carregando";
                if (!e.target.value) return symbols;

                return symbols.filter((item: { symbol: string }) =>
                  item.symbol
                    .toLocaleLowerCase()
                    .includes(e.target.value.toLocaleLowerCase())
                );
              });
            }}
          />
        </div>
        <StickyHeadTable
          rowsQntyPerPage={3000}
          noPagination
          tableWidth="350px"
          columns={columns1 as Column[]}
          rows={symbolsState}
          onClickRow={(rowId) => handleSymbolRowClick(rowId)}
        />
      </div>

      <StickyHeadTable
        columns={columns}
        rows={
          (bidData && [
            {
              last_price: bidData.data.P,
              ask_price: bidData.data.a,
              bid_price: bidData.data.b,
              price_change: bidData.data.P,
              symbol: bidData.data.s,
            },
          ]) ||
          []
        }
      />
    </S.Wrapper>
  );
};
