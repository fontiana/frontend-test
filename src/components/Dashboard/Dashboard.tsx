import React, { useEffect, useState } from 'react';
import { useSymbolContext } from '../../context/context';
import { ISymbol } from '../../models/symbol';
import * as S from './styles';

const Dashboard: React.FC = () => {
  const { symbolContext, symbolsList, setSymbolsList } = useSymbolContext();
  const [symbolData, setSymbolData] = useState<ISymbol[]>([]);
  const [listNames, setListNames] = useState<string[]>([]);
  const [isDropdownOpen, setIsDropdownOpen] = useState(false);

  const handleSelectChange = () => {
    setIsDropdownOpen(!isDropdownOpen);
  };

  const BASE_URL = `wss://stream.binance.com:9443/stream?streams=`;

  const url = BASE_URL + symbolContext.map((symbol) =>
    `${symbol.symbol.toLowerCase()}@ticker`).join('/')

  useEffect(() => {
    const socket = new WebSocket(url);
    socket.onmessage = (event) => {
      const { data } = JSON.parse(event.data);
      const newSymbolData = {
        symbol: data.s,
        lastPrice: data.c,
        bestBidPrice: data.b,
        bestAskPrice: data.a,
        priceChangePercent: data.P
      };

      setSymbolData((prevData) => {
        const updatedData = [...prevData];
        const existingSymbolIndex = updatedData.findIndex((item) => item.symbol === data.s);
        if (existingSymbolIndex !== -1) {
          updatedData[existingSymbolIndex] = newSymbolData;
        } else {
          updatedData.push(newSymbolData);
        }
        return updatedData;
      });
    }

    const names = symbolContext.map((_, index) => `List ${index + 1}`);
    setListNames(names);
  }, [url, symbolContext]);

  const handleListSelection = (listName: string) => {
    const selectedListIndex = listNames.indexOf(listName);
    const selectedList = symbolsList[selectedListIndex];
    setSymbolsList(selectedList);
  };

  return (
    <div>
      <S.DropdownContainer>
        {isDropdownOpen ? (
          <S.SelectActive onChange={(e) => handleListSelection(e.target.value)}>
            {symbolsList.map((index) => (
              <S.Option key={index} value={`List ${index + 1}`}>{`List ${index + 1}`}</S.Option>
            ))}
          </S.SelectActive>
        ) : (
          <S.Select onChange={(e) => handleListSelection(e.target.value)}>
            {symbolsList.map((list, index) => (
              <S.Option key={index} value={`List ${index + 1}`}>{`List ${index + 1}`}</S.Option>
            ))}
          </S.Select>
        )}
      </S.DropdownContainer>
      <S.CustomTable>
        <S.TableHead>
          <tr>
            <S.TableHeader>Symbol</S.TableHeader>
            <S.TableHeader>Last Price</S.TableHeader>
            <S.TableHeader>Bid Price</S.TableHeader>
            <S.TableHeader>Ask Price</S.TableHeader>
            <S.TableHeader>Price Change (%)</S.TableHeader>
          </tr>
        </S.TableHead>
        <tbody>
          {symbolData.map((symbol) => (
            <S.TableRow key={symbol.symbol}>
              <S.TableCell>{symbol.symbol}</S.TableCell>
              <S.TableCell>{symbol.lastPrice}</S.TableCell>
              <S.TableCell>{symbol.bestBidPrice}</S.TableCell>
              <S.TableCell>{symbol.bestAskPrice}</S.TableCell>
              <S.TableCell>{symbol.priceChangePercent}</S.TableCell>
            </S.TableRow>
          ))}
        </tbody>
      </S.CustomTable>
    </div>
  );
};

export default Dashboard;