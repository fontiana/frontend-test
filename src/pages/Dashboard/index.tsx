import { Suspense } from "react";
import ListSymbols from "./components/ListSymbols";
import Symbols from "./components/Symbols";
import { ExchangeInfoProvider } from "./context/useExchangeInfo";
import * as S from "./styles";
import Spinner from "../../components/Spinner";

const Dashboard = () => {
  return (
    <ExchangeInfoProvider>
      <S.Wrapper>
        <Suspense
          fallback={
            <S.Wrapper>
              <Spinner message="Loading symbols..." />
            </S.Wrapper>
          }
        >
          <Symbols />
        </Suspense>
        <ListSymbols />
      </S.Wrapper>
    </ExchangeInfoProvider>
  );
};

export default Dashboard;
