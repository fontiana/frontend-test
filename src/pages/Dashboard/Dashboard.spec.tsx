import Dashboard from ".";
import renderer from "react-test-renderer";

describe("Dashboard", () => {
  it("should render correctly", () => {
    const tree = renderer.create(<Dashboard />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
