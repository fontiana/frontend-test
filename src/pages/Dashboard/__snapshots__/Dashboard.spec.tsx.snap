// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dashboard should render correctly 1`] = `
<main
  className="sc-dJGLgI eNSlAv"
>
  <section
    className="sc-irLvoH KLyOf"
  >
    <section
      className="sc-fmKFmJ jalVDq"
      data-testid="undefined__spinner"
    >
      <div
        className="sc-fHekAb dsiVxd"
      />
      <h1>
        Loading symbols...
      </h1>
    </section>
  </section>
  <section
    className="sc-beyTiQ jeLTOA"
  >
    <section
      className="sc-guDMob gxxYoy"
    >
      <div
        className="sc-eDLKEg ilpPtz"
      >
        <label
          htmlFor="listOfExchanges"
        >
          Select a list:
        </label>
        <select
          className="sc-dmyDmy ifWSDO"
          id="listOfExchanges"
          onChange={[Function]}
          value=""
        />
        <span
          className="sc-hLQTFJ lliFRa"
          onClick={[Function]}
        >
          +
        </span>
      </div>
    </section>
    <table
      className="sc-iBdoyZ dvrDvo"
    >
      <thead
        className="sc-fsYfxw iwuMRD"
      >
        <tr>
          <th
            className="sc-qZruQ blByuq"
          >
            Symbol
          </th>
          <th
            className="sc-qZruQ blByuq"
          >
            Last Price
          </th>
          <th
            className="sc-qZruQ blByuq"
          >
            Bid Price
          </th>
          <th
            className="sc-qZruQ blByuq"
          >
            Ask Price
          </th>
          <th
            className="sc-qZruQ blByuq"
          >
            Price Change (%)
          </th>
        </tr>
      </thead>
      <tbody />
    </table>
  </section>
</main>
`;
